BatteryStateOfChargeService.cpp

```diff
@@ -0,0 +1,80 @@
+#include "BatteryStateOfChargeService.h"
+#include "BatteryData.h"
+
+namespace
+{
+    const double BATTERY_AMP_HOUR_CAPACITY = 123.0;
+}
+
+BatteryStateOfChargeService::BatteryStateOfChargeService(double initialStateOfChargePercent)
+: initialStateOfChargePercent_(initialStateOfChargePercent)
+{
+    ampHourUsed_ = 0.0;
+    timeTilCharged_ = QTime (0,0,0,0);
+    timeTilDepleted_ = QTime (0,0,0,0);
+}
+
+BatteryStateOfChargeService::~BatteryStateOfChargeService()
+{
+}
+
+double BatteryStateOfChargeService::totalAmpHoursUsed() const
+{
+    return ampHourUsed_;
+}
+
+bool BatteryStateOfChargeService::isCharging() const
+{
+    return isCharging_;
+}
+
+QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
+{
+   if(isCharging_)
+   {
+      return timeTilCharged_;
+   }
+
+   else
+   {
+      return timeTilDepleted_;
+   }
+}
+
+void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
+{
+    const double MILLISECONDS_TO_HOURS = 1/3600000.0;
+    const double PERCENT_TO_DECIMAL = 0.01;
+        if(ampHourUsed_ == 0.0)
+        {
+             ampHourUsed_ = BATTERY_AMP_HOUR_CAPACITY*initialStateOfChargePercent_*PERCENT_TO_DECIMAL;
+        }
+        else
+        {
+             ampHourUsed_ += ((previousCurrent_+batteryData.current)/2)*((previousTime_).msecsTo(batteryData.time)*MILLISECONDS_TO_HOURS);

```

Comment:
This one line is doing a lot, splitting it into multiple calculations would improve readability.